### PR TITLE
Gil/timing improvements

### PR DESCRIPTION
**Pull request by gilbran9**
This corrects velodyne point timestamps. It's not perfect but a large improvement already. It also modified the output point type to one that contains time. This point is also output from luminar too so that there's a standard point between the two. 

The firing timing is extracted from the master branch of this repo. 

The timing is adjusted from their solution because there's a bug in their code and they use the average of packet retrieval time as if it was the start of the packet retrieval process.

It should be merged at the same time as https://github.com/embarktrucks/brain/pull/3675

### CODE REVIEW

File: velodyne_driver/src/lib/input.cc [hunk 1]
```diff
@@ -199,11 +199,9 @@ namespace velodyne_driver
                          << nbytes << " bytes");
       }
 
-    // Average the times at which we begin and end reading.  Use that to
-    // estimate when the scan occurred. Add the time offset.
-    double time2 = ros::Time::now().toSec();
-    pkt->stamp = ros::Time((time2 + time1) / 2.0 + time_offset);
-
+    // Use ros time now as when we start getting packet as the time of scan. At this point
```
Comment (erik-embarktrucks) [inline velodyne_driver/src/lib/input.cc]: my brain hurts a bit from this sentence. Should it be:
"Use ros time now _from_ when we _started_ getting a packet as the time of _a_ scan"

Also, one scan contains many packets. This implies that one of them sets the timestamp for the whole scan, an explanation what happens downstream with the time would be nice.

File: velodyne_pointcloud/src/lib/rawdata.cc [hunk 2]
```diff
@@ -120,6 +124,35 @@ namespace velodyne_rawdata
...
+  std::vector<std::vector<double>> RawData::getVLP32TimingOffsets() {
+    // timing table calculation, from velodyne user manual
+
+    // 12 firings cycles in a data package, 32 lasers:
+    std::vector<std::vector<double>> timing_offsets(12, std::vector<double>(32, 0.));
+
+    // Time constants
+    const double full_firing_cycle = VLP_32_SPEC.firing_seq_duration * 1e-6;
+    const double single_firing = VLP_32_SPEC.firing_duration * 1e-6;
+    const bool dual_mode = false;
```
Comment (erik-embarktrucks) [inline velodyne_pointcloud/src/lib/rawdata.cc:138]: We can in theory set the VLP32 to use dual mode. We should support this (not hard code to false but take from config)

File: velodyne_pointcloud/src/lib/rawdata.cc [hunk 3]
```diff
@@ -458,14 +494,21 @@ namespace velodyne_rawdata
             intensity = (intensity > max_intensity) ? max_intensity : intensity;
 
             if (pointInRange(distance)) {
+              // Set point time as beginning of scan and then apply timing offset:
+              ros::Time pt_time = pkt.stamp;
+              if (!timing_offsets_.empty()) {
+                  pt_time = pt_time + ros::Duration(timing_offsets_[block][firing_seq]);
```
Comment (erik-embarktrucks) [inline velodyne_pointcloud/src/lib/rawdata.cc]: I'm wondering about the performance of constructor of `ros::Duration`. It's in the innermost place of this loop (that processes ~ 1 million points per second).

ros::Time is a subclass of some base class also, idk if there is a vtable lookup to access `sec` and `nsec` probably not (or it doesn't matter due to branch prediction). We should run intel vtunes for this node just to make sure there is nothing that surprises us in terms of performance.

**Review by erik-embarktrucks — CHANGES_REQUESTED** [review]
Hard to understand comment, then we should profile this node to make sure it's performant.

File: velodyne_pointcloud/src/lib/rawdata.cc [hunk 4]
```diff
@@ -270,15 +303,18 @@ namespace velodyne_rawdata
           intensity = (intensity > max_intensity) ? max_intensity : intensity;
 
           if (pointInRange(distance)) {
+            ros::Time pt_time = pkt.stamp; // No firing correction for this model
```
Comment (d1sounds) [inline velodyne_pointcloud/src/lib/rawdata.cc]: `const`

File: velodyne_pointcloud/src/lib/rawdata.cc (hunk 3 above)
Comment (d1sounds) [reply velodyne_pointcloud/src/lib/rawdata.cc]: Why not make your `timing_offsets_` be `vector<vector<Duration>>`?

Comment (d1sounds) [inline velodyne_pointcloud/src/lib/rawdata.cc]: what's the relationship between this time_offset patching and the one that happens in `getPacket()`?  Following this code, I initially assumed `timing_offsets_` values would get passed to `getPacket()`..

File: velodyne_driver/src/lib/input.cc [hunk 5]
```diff
@@ -114,7 +114,7 @@ namespace velodyne_driver
   /** @brief Get one velodyne packet. */
   int InputSocket::getPacket(velodyne_msgs::VelodynePacket *pkt, const double time_offset)
   {
-    double time1 = ros::Time::now().toSec();
+    const ros::Time time1 = ros::Time::now();
```
Comment (d1sounds) [inline velodyne_driver/src/lib/input.cc]: can you use a more meaningful variable name here?

File: velodyne_pointcloud/src/lib/rawdata.cc (hunk 3 above)
Comment (d1sounds) [reply velodyne_pointcloud/src/lib/rawdata.cc]: I think I'm missing the high level picture here.

It looks like all packets in a scan used to get timestamps with .. something (average between the scan start and the process time of each packet? or maybe the end of the scan?).

Now, we're using the start time of the scan... plus the computed time offset?

Can you fill in the gaps for me?

**Review by d1sounds — COMMENTED** [review]
Code looks ok, but I'm missing the high level of what this used to do and does now.

File: velodyne_pointcloud/src/lib/rawdata.cc (hunk 2 above)
Comment (gilbran9) [reply velodyne_pointcloud/src/lib/rawdata.cc:138]: The value is not in the config.

File: velodyne_pointcloud/src/lib/rawdata.cc (hunk 4 above)
Comment (gilbran9) [reply velodyne_pointcloud/src/lib/rawdata.cc]: It gets changed 2 lanes later...

File: velodyne_pointcloud/src/lib/rawdata.cc (hunk 3 above)
Comment (gilbran9) [reply velodyne_pointcloud/src/lib/rawdata.cc]: fixed.

Comment (gilbran9) [reply velodyne_pointcloud/src/lib/rawdata.cc]: Added comments to explain.

File: velodyne_driver/src/lib/input.cc (hunk 1 above)
Comment (gilbran9) [reply velodyne_driver/src/lib/input.cc]: fixed.

File: velodyne_pointcloud/src/lib/rawdata.cc (hunk 4 above)
Comment (gilbran9) [reply velodyne_pointcloud/src/lib/rawdata.cc]: fixed.

File: velodyne_driver/src/lib/input.cc (hunk 5 above)
Comment (gilbran9) [reply velodyne_driver/src/lib/input.cc]: fixed.

**Review by erik-embarktrucks — APPROVED** [review]

**Pull request merged**